### PR TITLE
#4091 - Update FuelType/EndUseFuelType (and Sql) to match E+ 9.4.0

### LINKED ISSUE #4091: Doesn't look like OpenStudio::EndUseFuelType.getValues was updated for EnergyPlus 9.4
See link to issue on common measure for OpenStudio Results. The old code would loop through fuel types convert integer to string and then pass it to SQL tabular query. Additional Fuel was removed and a number of specific additional fuels were added.
https://github.com/NREL/openstudio-common-measures-gem/issues/26#issuecomment-698000362

My work around in measure for now is to use hard coded list of fuel types instead of the method.

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes #4091
 - Update FuelType/EndUseFuelType (and Sql) to match E+ 9.4.0

Note: this includes #4085 (Update to v9.4.0-RC2)

### Pull Request Author

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Checked behavior in OpenStudioApplication, adjusted policies as needed (`src/openstudio_lib/library/OpenStudioPolicy.xml`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [x] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [x] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/utilities/data/DataEnums.hpp [hunk 1]
```diff
@@ -66,27 +66,33 @@ OPENSTUDIO_ENUM(FuelType,
...
+  ((Coal))
   ((FuelOil_1)(FuelOilNo1))
   ((FuelOil_2)(FuelOilNo2))
   ((Propane))
   ((Water))
   ((Steam))
   ((DistrictCooling))
   ((DistrictHeating))
+  ((OtherFuel_1)(OtherFuel1))
+  ((OtherFuel_2)(OtherFuel2))
```
Comment (jmarrec) [inline src/utilities/data/DataEnums.hpp:78]: FuelType gets new values `Coal` and `OtherFuel_1`/`OtherFuel_2`

File: src/utilities/data/DataEnums.hpp [hunk 2]
```diff
@@ -103,18 +109,32 @@ typedef std::vector<FuelType> FuelTypeVector;
...
+  ((Coal))
+  ((FuelOil_1)(Fuel Oil No 1))
+  ((FuelOil_2)(Fuel Oil No 2))
+  ((Propane))
+  ((OtherFuel_1)(Other Fuel 1))
+  ((OtherFuel_2)(Other Fuel 2))
   ((DistrictCooling)(District Cooling))
   ((DistrictHeating)(District Heating))
-  ((Water)(Water)));
+  ((Water)));
```
Comment (jmarrec) [inline src/utilities/data/DataEnums.hpp:137]: EndUseFuelType gets a bunch of new values.

```
SELECT DISTINCT(ColumnName) FROM TabularDataWithStrings
WHERE ReportName="AnnualBuildingUtilityPerformanceSummary"
AND TableName="End Uses"
```

```

Electricity
Natural Gas
Gasoline
Diesel
Coal
Fuel Oil No 1
Fuel Oil No 2
Propane
Other Fuel 1
Other Fuel 2
District Cooling
District Heating
Water
```

File: src/utilities/sql/SqlFile_Impl.cpp [hunk 3]
```diff
@@ -1005,40 +1005,7 @@ namespace openstudio{
...
-        else if (fuel == FuelType::Propane){
-          meterName = "PROPANE:FACILITY";
-        }
-        else if (fuel == FuelType::Steam){
-          meterName = "STEAM:FACILITY";
-        }
-        else if (fuel == FuelType::EnergyTransfer){
-          meterName = "ENERGYTRANSFER:FACILITY";
-        }
+        std::string meterName = boost::to_upper_copy(fuel.valueDescription()) + ":FACILITY";
```
Comment (jmarrec) [inline src/utilities/sql/SqlFile_Impl.cpp:1008]: Avoid hardcoding stuff. Use `valueDescription` (second parenthesis) from ENUM instead.

File: src/utilities/sql/SqlFile_Impl.cpp [hunk 4]
```diff
@@ -1093,19 +1060,12 @@ namespace openstudio{
...
-
       // Loop through all fuels and add up their costs
-      for (const auto& fuelType : fuelTypes) {
+      for (int i : openstudio::FuelType::getValues()) {
+        openstudio::FuelType fuelType(i);
+
         // Get the annual energy cost
-        boost::optional<double> cost = annualTotalCost(fuelType);
-        if (cost) {
+        if (boost::optional<double> cost = annualTotalCost(fuelType)) {
```
Comment (jmarrec) [inline src/utilities/sql/SqlFile_Impl.cpp:1068]: Same, avoid hardcoding it, so we can just add/modify the ENUM instead

File: src/sdd/ReverseTranslator.cpp [hunk 5]
```diff
@@ -910,7 +910,10 @@ namespace sdd {
           fuelType == FuelType::Propane ||
           fuelType == FuelType::Water ||
           fuelType == FuelType::Steam ||
+          fuelType == FuelType::OtherFuel_1 ||
+          fuelType == FuelType::OtherFuel_2 ||
           fuelType == FuelType::EnergyTransfer){
+        // TODO: why are we ignoring these?! Aside from EnergyTransfer at least
```
Comment (jmarrec) [inline src/sdd/ReverseTranslator.cpp:916]: I don't really follow this part, so left a TODO.

File: src/model/UtilityBill.cpp [hunk 6]
```diff
@@ -594,14 +594,14 @@ namespace detail {
...
       //  result.push_back("tons");
       //  result.push_back("lbs");
       //  result.push_back("kLbs");
       //  result.push_back("MLbs");
-      //  result.push_back("kBtu");
-      //  result.push_back("MBtu");
-      //  break;
+        result.push_back("kBtu");
+        result.push_back("MBtu");
+        break;
```
Comment (jmarrec) [inline src/model/UtilityBill.cpp:604]: In UtilityBill, support Coal. But do not add weight-absed units as conversion isn't fixed, it depends on the type of coal.

File: src/model/UtilityBill.cpp [hunk 7]
```diff
@@ -647,6 +647,14 @@ namespace detail {
...
         result.push_back("MBtu");
         break;
+      case FuelType::OtherFuel_1:
+        result.push_back("kBtu");
+        result.push_back("MBtu");
+        break;
+      case FuelType::OtherFuel_2:
+        result.push_back("kBtu");
+        result.push_back("MBtu");
+        break;
```
Comment (jmarrec) [inline src/model/UtilityBill.cpp:657]: Add OtherFuel1&2 to UtilityBill too.

File: src/model/OutputMeter.cpp [hunk 8]
```diff
@@ -467,7 +467,7 @@ IddObjectType OutputMeter::iddObjectType() {
 boost::regex OutputMeter::meterRegex()
 {
   // DLM: Must put more specific terms, e.g. HeatingCoils, before less specific terms, e.g. Heating
-  static const boost::regex result("^(.*?)?:?(InteriorLights|ExteriorLights|InteriorEquipment|ExteriorEquipment|Fans|Pumps|HeatingCoils|Heating|CoolingCoils|Cooling|HeatRejection|Humidifier|HeatRecoveryForCooling|HeatRecoveryForHeating|HeatRecovery|WaterSystems|Cogeneration|Refrigeration|Chillers|Boilers|Baseboard)?:?(Electricity|Gasoline|NaturalGas|Diesel|Coal|FuelOilNo1|FuelOilNo2|Propane|Water|Steam|DistrictCooling|DistrictHeating|EnergyTransfer)?:?(Facility|Building|HVAC|Zone|System|Plant)?:?([^:]*?)?$");
+  static const boost::regex result("^(.*?)?:?(InteriorLights|ExteriorLights|InteriorEquipment|ExteriorEquipment|Fans|Pumps|HeatingCoils|Heating|CoolingCoils|Cooling|HeatRejection|Humidifier|HeatRecoveryForCooling|HeatRecoveryForHeating|HeatRecovery|WaterSystems|Cogeneration|Refrigeration|Chillers|Boilers|Baseboard)?:?(Electricity|Gasoline|NaturalGas|Diesel|Coal|FuelOilNo1|FuelOilNo2|Propane|Water|Steam|DistrictCooling|DistrictHeating|OtherFuel1|OtherFuel2|EnergyTransfer)?:?(Facility|Building|HVAC|Zone|System|Plant)?:?([^:]*?)?$");
```
Comment (jmarrec) [inline src/model/OutputMeter.cpp:470]: Needed to update the meterRegex to support OtherFuel1|OtherFuel2

File: src/model/LifeCycleCostParameters_Impl.hpp [hunk 9]
```diff
@@ -164,6 +164,10 @@ namespace detail {
 
     boost::optional<double> waterInflation() const;
 
+    boost::optional<double> otherFuel1Inflation() const;
+
+    boost::optional<double> otherFuel2Inflation() const;
+
```
Comment (jmarrec) [inline src/model/LifeCycleCostParameters_Impl.hpp:170]: Add otherFuelXInflation to LifeCycleCostParameters.

File: src/energyplus/ForwardTranslator/ForwardTranslateLifeCycleCostParameters.cpp [hunk 10]
```diff
@@ -165,6 +165,8 @@ boost::optional<IdfObject> ForwardTranslator::translateLifeCycleCostParameters(
     makeUsePriceEscalation(modelObject.fuelOil2Inflation(),    "FuelOilNo2", baseDateMonth, baseDateYear, lengthOfStudyPeriodInYears, m_idfObjects);
     makeUsePriceEscalation(modelObject.propaneInflation(),     "Propane", baseDateMonth, baseDateYear, lengthOfStudyPeriodInYears, m_idfObjects);
     makeUsePriceEscalation(modelObject.waterInflation(),       "Water", baseDateMonth, baseDateYear, lengthOfStudyPeriodInYears, m_idfObjects);
+    makeUsePriceEscalation(modelObject.otherFuel1Inflation(),  "OtherFuel1", baseDateMonth, baseDateYear, lengthOfStudyPeriodInYears, m_idfObjects);
+    makeUsePriceEscalation(modelObject.otherFuel2Inflation(),  "OtherFuel2", baseDateMonth, baseDateYear, lengthOfStudyPeriodInYears, m_idfObjects);
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslateLifeCycleCostParameters.cpp:169]: Handle OtherFuel1/2 in the FT for LifeCycleCostParameters

File: src/energyplus/EnergyPlus.qrc [hunk 11]
```diff
@@ -1,18 +0,0 @@
-<!DOCTYPE RCC><RCC version="1.0">
```
Comment (jmarrec) [inline src/energyplus/EnergyPlus.qrc:1]: This file was a leftover from the App separation it seems.

File: resources/model/OpenStudio.idd [hunk 12]
```diff
@@ -30712,7 +30712,17 @@ OS:LifeCycleCost:Parameters,
...
+       \note Applies only if Analysis Use NIST Fuel Escalation Rates is No
+       \note Enter the rate of inflation for water as a decimal. For a 2% rate enter
+       \note the value 0.02.
+       \type real
+  N18, \field OtherFuel1 Inflation
+       \note Applies only if Analysis Use NIST Fuel Escalation Rates is No
+       \note Enter the rate of inflation for water as a decimal. For a 2% rate enter
+       \note the value 0.02.
+       \type real
+  N19; \field OtherFuel2 Inflation
```
Comment (jmarrec) [inline resources/model/OpenStudio.idd:30725]: New fields for IDD in `OS:LifeCycleCost:Parameters`

File: resources/model/OpenStudio.idd [hunk 13]
```diff
@@ -30796,13 +30806,16 @@ OS:UtilityBill,
...
+       \key Coal
        \key FuelOil_1
        \key FuelOil_2
        \key Propane
        \key Water
        \key Steam
        \key DistrictCooling
        \key DistrictHeating
+       \key OtherFuel_1
+       \key OtherFuel_2
```
Comment (jmarrec) [inline resources/model/OpenStudio.idd:30818]: OS:UtilityBill: add `Coal` and `OtherFuel1/2`

**Comment by jmarrec** [comment]
I think I have a working implementation. Need review and testing now @DavidGoldwasser @kbenne

**Comment by tijcolem** [comment]
So it looks like we need to :
Merge - https://github.com/NREL/OpenStudio-workflow-gem/pull/120 
Release new workflow-gem v 2.0.2
Update Openstudio-gems for OpenStudio with new workflow
Merge this PR

**Review by DavidGoldwasser — APPROVED** [review]
This looks good. I'm able to use the fuel types listed in OpenStudio end use fuels instead of hard coding them. I still have issue getting monthly gas end uses for monthly overview and HVAC load profiles, but that is different issue which should be updated with updated workflow gem.

**Comment by jmarrec** [comment]
@tijcolem I added an actual checklist to your comment https://github.com/NREL/OpenStudio/pull/4093#issuecomment-698597793 and set the status. I do not know how to pick up the openstudio gems tar.gz that jenkins built for https://github.com/NREL/openstudio-gems/pull/31 though.

**Comment by jmarrec** [comment]
Removed "Pull Request - Ready for CI" label since it'll be built in #4100

**Pull request merged**